File: main/ctci-merge-sort/ctci-merge-sort.py
#!/usr/bin/env python3

def count_swaps(a):
    '''Sort a by mergesort and return the number of swaps.'''
    aux = [None] * len(a)

    def merge(lo, hi):
        mid = (lo + hi) // 2

        i = lo
        j = mid + 1
        k = lo

        swaps = 0
        while i <= mid and j <= hi: # merge from left and right sides
            if a[i] <= a[j]:
                aux[k] = a[i]
                i += 1
            else:
                swaps += j - k
                aux[k] = a[j]
                j += 1
            k += 1

        while i <= mid:     # merge any remaining from left
            aux[k] = a[i]
            i += 1
            k += 1

        while j <= hi:      # ...or any remaining from right
            aux[k] = a[j]
            j += 1
            k += 1

        while lo != k:      # copy back after merge
            a[lo] = aux[lo]
            lo += 1

        return swaps

    def mergesort(lo, hi):
        if lo >= hi:
            return 0

        mid = (lo + hi) // 2
        swaps = mergesort(lo, mid) + mergesort(mid + 1, hi)
        return swaps + merge(lo, hi)

    return mergesort(0, len(a) - 1)

for _ in range(int(input())):
    input() # don't need n
    print(count_swaps(list(map(int, input().split()))))

File: main/ctci-merge-sort/ctci-merge-sort.cs
using System;

[assembly: System.Reflection.AssemblyVersion("1.0.0.0")]

internal static class Algo {
    private static long Merge(this int[] a, int[] aux, int lo, int hi)
    {
        var mid = lo + (hi - lo) / 2;

        var i = lo;
        var j = mid + 1;
        var k = lo;
        
        var swaps = 0L;
        while (i <= mid && j <= hi) { // merge from left and right sides
            if (a[i] <= a[j]) aux[k++] = a[i++];
            else {
                swaps += j - k;
                aux[k++] = a[j++];
            }
        }

        while (i <= mid) aux[k++] = a[i++]; // merge any remaining from left
        while (j <= hi)  aux[k++] = a[j++]; // ...or any remaining from right

        for (; lo != k; ++lo) a[lo] = aux[lo]; // copy back after merge
        return swaps;
    }

    private static long Mergesort(this int[] a, int[] aux, int lo, int hi)
    {
        if (lo >= hi) return 0;

        var mid = lo + (hi - lo) / 2;
        var swaps = a.Mergesort(aux, lo, mid) + a.Mergesort(aux, mid + 1, hi);
        return swaps + a.Merge(aux, lo, hi);
    }

    internal static long Mergesort(this int[] a) // returns number of swaps
        => a.Mergesort(new int[a.Length], 0, a.Length - 1);
}

internal static class Solution {
    private static void Main()
    {
        for (var d = int.Parse(Console.ReadLine()); d > 0; --d) {
            Console.ReadLine(); // don't need n
            var a = Array.ConvertAll(Console.ReadLine().Split(), int.Parse);
            Console.WriteLine(a.Mergesort());
        }
    }
}
